distributed/tests/test_worker_state_machine.py

```diff
@@ -1002,6 +1005,195 @@ def test_gather_priority(ws):
     ]
 
 
+def test_constrained_task_handles_resources_on_success(ws_with_running_task):
+    ws = ws_with_running_task
+    assert ws.available_resources == {"R": 0}
+    # Ensure that x transitions through the ``constrained`` state
+    assert_story(
+        ws.story("x"),
+        [
+            ("x", "compute-task", "released"),
+            ("x", "released", "waiting", "waiting", {"x": "ready"}),
+            ("x", "waiting", "ready", "waiting", {"x": "constrained"}),
+            ("x", "waiting", "constrained", "constrained", {"x": "executing"}),
+            ("x", "constrained", "executing", "executing", {}),
+        ],
+        strict=False,
+    )
```

Comment:
As a general design note, we'd rather avoid using WorkerState.story as it is a level of abstraction that is well below that of events.
```suggestion
```

Reply:
Sounds good, I wanted to ensure that we go through the "constrained" state in all cases, but I guess testing for resource allocation is the only thing we're trying to achieve in the end.